File: src/main/java/com/uiptv/util/StalkerAttributeParser.java
package com.uiptv.util;

import java.text.Normalizer;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.regex.Pattern;

/**
 * Interface for attribute parsers used by the StalkerPortalParser.
 */
public interface StalkerAttributeParser {
    
    /**
     * Parses the line and returns the extracted value, or null if not found.
     */
    String parse(String line);

    /**
     * Returns the type of attribute this parser handles.
     */
    StalkerAttributeType getAttributeType();

    default boolean lineContainsKeyword(String line, List<String> keywords) {
        String normalizedLine = Normalizer.normalize(line, Normalizer.Form.NFKC).toLowerCase();
        return keywords.stream().anyMatch(normalizedLine::contains);
    }

    default String extractHexValueFromLine(String line) {
        String normalized = Normalizer.normalize(line, Normalizer.Form.NFKC);
        String cleaned = normalized.replaceAll("[^\\x20-\\x7E]", " ").trim();
        List<String> tokens = new ArrayList<>(List.of(cleaned.split("\\s+")));
        Collections.reverse(tokens);
        for (String token : tokens) {
            if (Pattern.compile("^[0-9A-Fa-f]{6,}$").matcher(token).matches()) {
                return token;
            }
        }
        return null;
    }
}

File: server/src/main/kotlin/com/uiptv/api/StalkerAttributeParser.kt
package com.uiptv.api

import com.uiptv.util.StalkerAttributeType
import java.text.Normalizer
import java.util.regex.Pattern

interface StalkerAttributeParser {
    fun parse(line: String): String?

    fun getAttributeType(): StalkerAttributeType

    fun lineContainsKeyword(line: String, keywords: List<String>): Boolean {
        val normalizedLine = Normalizer.normalize(line, Normalizer.Form.NFKC).lowercase()
        return keywords.any(normalizedLine::contains)
    }

    fun extractHexValueFromLine(line: String): String? {
        val normalized = Normalizer.normalize(line, Normalizer.Form.NFKC)
        val cleaned = normalized.replace(Regex("[^\\x20-\\x7E]"), " ").trim()
        val tokens = cleaned.split(Regex("\\s+"))
        for (token in tokens.asReversed()) {
            if (Pattern.compile("^[0-9A-Fa-f]{6,}$").matcher(token).matches()) {
                return token
            }
        }
        return null
    }
}
